File: LC/[0947] Most Stones Removed with Same Row or Column.cpp
class Solution {
    unordered_map<int, int> Parent;
    unordered_map<int, vector<int>> MapX;
    unordered_map<int, vector<int>> MapY;
    int N = 10000;
public:
    int removeStones(vector<vector<int>>& stones) {
        for(auto x: stones){
            int ii=x[0];
            int jj=x[1];
            int gid = ii*N+jj;
            Parent[gid] = gid;
            MapX[ii].push_back(gid);
            MapY[jj].push_back(gid);
        }

        for(auto x: MapX){
            int id0 = x.second[0];
            for(int ii=1; ii<x.second.size(); ++ii){
                int id = x.second[ii];
                if(Find(id0) != Find(id)){
                    Union(id0, id);
                }
            }
        }

        for(auto x: MapY){
            int id0 = x.second[0];
            for(int ii=1; ii<x.second.size(); ++ii){
                int id = x.second[ii];
                if(Find(id0) != Find(id)){
                    Union(id0, id);
                }
            }
        }

        unordered_set<int> Set;
        for(auto x: stones){
            int id = x[0]*N+x[1];
            int id0=Find(id);
            Set.insert(id0);
        }

        return stones.size() - Set.size();
    }

    int Find(int x){
        if(Parent[x] != x) Parent[x] = Find(Parent[x]);
        return Parent[x];
    }

    void Union(int a, int b){
        int pa = Find(a);
        int pb = Find(b);
        if(pa < pb)
            Parent[pb] = pa;
        else    
            Parent[pa] = pb;
    }
};

File: LC/[0947] Most Stones Removed with Same Row or Column.py
class DSU:
    def __init__(self):
        self.parent = {}
        self.size = {}

    def find(self, x):
        if x not in self.parent:
            self.parent[x] = x
            self.size[x] = 1

        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])

        return self.parent[x]

    def union(self, a, b):
        pa = self.find(a)
        pb = self.find(b)

        if pa == pb:
            return

        if self.size[pa] < self.size[pb]:
            pa, pb = pb, pa

        self.parent[pb] = pa
        self.size[pa] += self.size[pb]


class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        dsu = DSU()

        for x, y in stones:
            row_id = ("row", x)
            col_id = ("col", y)
            dsu.union(row_id, col_id)

        roots = set()
        for x, y in stones:
            roots.add(dsu.find(("row", x)))

        return len(stones) - len(roots)

# ## Strategy
# * Treat each row and each column as DSU nodes:
#   row node = ('row', x), column node = ('col', y)
# * A stone at (x, y) connects row x and column y, so union(('row', x), ('col', y)).
# * If another stone shares the same row or column, it will connect through the same row/col node.
# * Therefore, each connected component represents one group of stones.
# * In each group, we can remove all stones except one.
# * Answer = total stones - number of connected components.